### ISSUE TITLE
[Question] Evaluating the loss for Autoencoders

### ISSUE THREAD

**Issue by pi-tau** [phlippe/uvadlc_notebooks#108]
**Tutorial**: 9

Hi @phlippe 
I am trying to implement a Variational Autoencoder and I was using this tutorial as a helper.
My question is regarding the calculation of the loss value. In Cell 6 from Tutorial 9 you calculate the reconstruction loss by summing over the input dimensions and averaging only over the batch. Using this approach I obtain very large values for the loss, and consequently for the gradients (grad_norm ~ 500) and I am not sure if that is ok. It still manages to reduce the loss and produces blurry but colorful images for cifar10, and somewhat good images of digits for mnist. But the gradient norm seems very large. I tried also using gradient clipping (clip value of 10.) but then the gradient is clipped at every single update step, and that does not seem like the way to go.

Could you please report the size of the grad norms during training the Autoencoder and what is your opinion on these values ? Is a value of grad norm ~ 500 too much or is it ok?
Maybe you could also add a plot showing the training loss per iteration (maybe just for latent_dim=256) ?

I also tried computing the loss by taking the mean over all dimensions, but then the grad norm is ~ 0.1 and now this seems to low and I am not sure in what order should the learning rate be (0.1?, 1e-2?).

Finally, the variational autoencoder models the loss as `total_loss = recon_loss + KL_loss` where the KL loss is the KL divergence between the encoder distribution and the prior: $KLLoss = KL(q(z|x) || p_Z(z))$. In this case the `recon_loss` has dimension equal to the original space (e.g. `3x32x32`) while the `KL_loss` has dimension equal to the latent space (e.g. `128`), and I think that computing the total loss as: `total_loss = recon_loss.mean(dim=0).sum() + KL_loss.mean(dim=0).sum()` is the only correct way. I think that computing it as `total_loss = recon_loss.mean() + KL_loss.mean()` would actually scale the two losses differently and one might dominate the other. Is that correct or am I missing something ?

Any info would be greatly appreciated. Thanks :)

**Comment by phlippe** [comment]
Hi @pi-tau, when you use Adam, the gradients are scaled back to a unit norm, such that multiplying the loss by a constant does not make a difference. This is the case here for either taking the mean or the sum. 
For a VAE, you indeed have to sum the reconstruction loss over pixels to use the ELBO as objective.

**Comment by pi-tau** [comment]
Hi, thanks for the useful info. I didn't know that.
In this case, if training with SGD or SGD+Momentum, would you simply clip the grad norm to 1. ?

**Comment by phlippe** [comment]
If you train with SGD, then you either have to adjust your learning rate and grad norm, or scale the loss back to typical values, e.g. by taking the mean instead of the sum.
